### PR TITLE
update the L1 and HLT post-LS1 customisations

### PR DESCRIPTION
**Pull request by fwyzard**
- add customisations to use the 25ns v0 and 50ns v0 L1 menus
- separate the L1 customisations for 50ns, 25ns and heavy ions
- move L1 HI menu customisation from L1Trigger_custom.py to customise_overwriteL1Menu.py
- update SLHCUpgradeSimulations/Configuration/python/postLS1Customs.py

### CODE REVIEW

**Comment by davidlange6** [comment]
@cmsbuild, please test

File: SLHCUpgradeSimulations/Configuration/python/postLS1Customs.py [hunk 1]
```diff
@@ -90,145 +88,8 @@ def customiseRun2EraExtras(process):
         process.simCsctfTrackDigis.SectorReceiverInput = cms.untracked.InputTag("simCscTriggerPrimitiveDigis", "MPCSORTED")
 
     # deal with L1 Emulation separately:
-    # replace the L1 menu from the global tag with one of the following alternatives
```
Comment (fwyzard) [inline SLHCUpgradeSimulations/Configuration/python/postLS1Customs.py]: Note that here I've replaced the copy of the L1 customisations with a call to the original customisation function: the copy had become out-of-sync from the original, and I hardly see the point of maintaining the same customisation in two places.

File: SLHCUpgradeSimulations/Configuration/python/postLS1Customs.py [hunk 2]
```diff
@@ -1,19 +1,17 @@
...
 def customisePostLS1(process):
 
     # deal with CSC separately:
+    from SLHCUpgradeSimulations.Configuration.muonCustoms import customise_csc_PostLS1
     process = customise_csc_PostLS1(process)
 
     # deal with L1 Emulation separately:
-    customiseSimL1EmulatorForPostLS1(process)
+    from L1Trigger.L1TCommon.customsPostLS1 import customiseSimL1EmulatorForPostLS1_25ns
+    process = customiseSimL1EmulatorForPostLS1_25ns(process)
```
Comment (fwyzard) [inline SLHCUpgradeSimulations/Configuration/python/postLS1Customs.py]: Note that I'm still calling the 25ns customisation unconditionally.

File: L1Trigger/L1TCommon/python/customsPostLS1.py [hunk 3]
```diff
@@ -135,30 +107,29 @@ def customiseSimL1EmulatorForPostLS1_nomenu(process):
...
-#        )
-#    print process.HLTSchedule
+
+    return process
+
+
+from L1Trigger.Configuration.customise_overwriteL1Menu import *
+
+def customiseSimL1EmulatorForPostLS1_50ns(process):
+    # move to the 50ns v0 menu once the HLT has been updated accordingly
```
Comment (fwyzard) [inline L1Trigger/L1TCommon/python/customsPostLS1.py]: Note that I'm still using the 25ns v2 L1 menu, as the HLT is not yet ready to use the dedicated 50ns v0 menu.

**Comment by fwyzard** [comment]
@davidlange6 , so far here there should be no changes to any workflow.

Should I make a copy of postLS1Customs.py specifically for 50ns, to load the different L1 menu ?
What about Heavy Ions ?

**Comment by fwyzard** [comment]
OK, I've added dedicated post-LS1 customisations for 25ns, 50ns and Heavy Ions.

However, they are not used yet: somebody that knows how would need to update Configuration/PyReleaseValidation/python/relval_steps.py to use them as appropriate.

**Comment by davidlange6** [comment]
we can handle this internally to the python instead. I'll take care

**Comment by civanch** [comment]
@fwyzard , @davidlange6, in the integrated PR #7503 HF shower library was moved from postLS1Customs.py to g4SimHits_cff.py and enabled via run2 era.

Is it optimal to have HF shower library customisation in 2 new py files for 25 and 50 ns? I mean should we keep or remove lines:
 
def customise_Sim(process):
   # enable 2015 HF shower library
   process.g4SimHits.HFShowerLibrary.FileName = 'SimG4CMS/Calo/data/HFShowerLibrary_npmt_noatt_eta4_16en_v3.root'
   return process

**Comment by davidlange6** [comment]
@cmsbuild, please test

**Comment by davidlange6** [comment]
Hi Vladimir
I believe you have not read the code completely correctly and instead see changes related to starting the support for eras.

David

**Comment by fwyzard** [comment]
Mhm, this is unexpected... I can double check tomorrow, but maybe the errors are unrelated ?

**Comment by Martin-Grunewald** [comment]
Hi
```
OK, I've added dedicated post-LS1 customisations for 25ns, 50ns and Heavy Ions.
```
that looks like a horrible duplication of (nearly) identical PostLS1 files.

**Comment by Martin-Grunewald** [comment]
Also, I still think that the three SLHCUpgradeSimulations/Configuration/python/postLS1Customs*
files for HI/25/50 have a huge overlap which may go out of sync... Ideally one would use
a single common one called by the three specific ones in addition to the three sets of
specific changes...

**Comment by fwyzard** [comment]
I agree, but haven't had the time to merge them properly yet.

Ciao,
.Andrea

**Comment by davidlange6** [comment]
As this stands, this PR loads the L1 menus for all workflows whether or not they have anything to do with L1. I believe the steps that need customiseSimL1EmulatorForPostLS1_25ns/50ns are "L1", "HLT", and "L1Reco". Is that correct?

**Comment by fwyzard** [comment]
Hi David,
you need the L1 menu also in DQM, and PAT, otherwise you read the wrong one
from the global tag.

I do not know if other workflows could be affected, too. Maybe AlCa ?

Ciao,
.Andrea

**Comment by davidlange6** [comment]
as an example, remind me how PAT uses the L1 menu directly? That would give me a good example to work from

**Comment by fwyzard** [comment]
Hi David,
I don't know by heart, I remember noticing that the PAT step accesses the
L1 menu, while investigating which step uses which conditions.

I guess it's to map the L1 trigger bits to trigger names.

Ciao,
.Andrea
as an example, remind me how PAT uses the L1 menu directly? That would give
me a good example to work from

On Feb 6, 2015, at 6:27 PM, Andrea Bocci <notifications@github.com>
wrote:

> Hi David,
> you need the L1 menu also in DQM, and PAT, otherwise you read the wrong
one
> from the global tag.
>
> I do not know if other workflows could be affected, too. Maybe AlCa ?
>
> Ciao,
> .Andrea
> On 6 Feb 2015 18:19, "David Lange" <notifications@github.com> wrote:
>
> > As this stands, this PR loads the L1 menus for all workflows whether or
> > not they have anything to do with L1. I believe the steps that need
> > customiseSimL1EmulatorForPostLS1_25ns/50ns are "L1", "HLT", and
"L1Reco".
> > Is that correct?
> >
> > �
> > Reply to this email directly or view it on GitHub
> > <https://github.com/cms-sw/cmssw/pull/7579#issuecomment-73275731>.
> >
> �
> Reply to this email directly or view it on GitHub.
>

—
Reply to this email directly or view it on GitHub
<https://github.com/cms-sw/cmssw/pull/7579#issuecomment-73280341>.

**Comment by davidlange6** [comment]
Ok - i'll see what I can learn.

**Comment by Martin-Grunewald** [comment]
PAT trigger code uses the HLTConfigProvider and also L1GtUtils to access L1
names which needs the proper L1 menu.
Any code using the HLTConfigProvider (DQM etc.) may access
L1 menu information incl. L1 trigger names.

**Comment by davidlange6** [comment]
ie, all cms analysis workflows possibly need the menu. ok - it needs to be back in the GT asap� not sure how to support this in the meanwhile, but will find something.

**Comment by fwyzard** [comment]
Rebased, and cleaned up the different cutsomisations.

This is my best bet at automatically picking up the correct customisation for 50ns vs. 25ns - but will likely fail for workflows that run the L1 emulator without running the mixing module.

Heavy Ions needs to be enabled explicitly.

Actually, the best solution would be to enable any of the three specific scenarios explicitly via cmsDriver.

**Comment by fwyzard** [comment]
@cmsbuild please test

**Comment by davidlange6** [comment]
I would suggest we just key the menu off of the global tag instead of trying to figure out the right pileup config for every workflow. This is in any case where the menus need to get back to.

**Comment by fwyzard** [comment]
- rebased
  - reverted the heavy ions L1 menu to the 25ns menu for the moment (until the relevant HLT is updated)
  - propagate the alternative customisePostLS1 to the matrix, leaving the 25ns case as the default

**Comment by fwyzard** [comment]
@cmsbuld please test

**Comment by fwyzard** [comment]
`runTheMatrix.py -l limited` seems to work for me, on top of CMSSW_7_4_X_2015-02-09-1400

**Comment by franzoni** [comment]
hello, I've taken a first look and intend to digest further.

Are we sure we want to go down this path ?
The complexity of interplay between the underlying 25ns default and the "tweak on top" actions is beyond what we can reliably control.  Kudos to @fwyzard et all who have put energies into this. 
I feel uncomfortable and that we're cooking up an unmaintainable system, am I the only one.

...when postLS1Customs was introduced in 2013 and put in production and relval,  the reassuring comment to those who pointed out it was not a good direction to go was: that it would be temporary and soon replaced by a more durable solution. :-) 

Before scrutinising the details of the relval_step I'll want to think more.

**Comment by fwyzard** [comment]
Hi, FWIW I agree with Giovanni about the maintainability of this approach.

Here, I simply wanted to make the 50ns/25ns/HI cases explicit in the matrix in order to allow different L1 emulator configuration and menu. The splitting of the post-LS1 customisation is kind of a side effect.

Is there a way to expand the python files for all the workflows in the matrix ?
That could be a way to make sure that no (unwanted) changes enter any workflow.

**Comment by davidlange6** [comment]
On Feb 10, 2015, at 10:17 AM, Andrea Bocci <notifications@github.com>
 wrote:

> Hi, FWIW I agree with Giovanni about the maintainability of this approach.
> 
> Here, I simply wanted to make the 50ns/25ns/HI cases explicit in the matrix in order to allow different L1 emulator configuration and menu. The splitting of the post-LS1 customisation is kind of a side effect.
> 
> Is there a way to expand the python files for all the workflows in the matrix ?
> That could be a way to make sure that no (unwanted) changes enter any workflow.
> 

you can call dumpPython on the config that gets created (e.g., file.write(process.dumpPython) )

**Comment by davidlange6** [comment]
Lets be clear that the complication here is that the "right" L1 menu can not be deduced simply in all cases, as it is needed by essentially all CMS workflows. This means that short of a user specifying that her data is 25ns or 50ns, there is no good solution aside from a GT driven one (so "eras" will suffer from the same need to specify enough information for 25/50) or by file providence (which means have an input file accessible). Other use cases are simpler as they can be derived from knowing that a "run 2" configuration is wanted and from the process PSet itself. 

The menu has been in the GT and people have written code assuming that it was in the GT - the current discussion is driven by the (hopefully) temporary move of the menu out of the DB.

**Comment by civanch** [comment]
+1

**Comment by fwyzard** [comment]
On 10 February 2015 at 10:51, David Lange <notifications@github.com> wrote:

> you can call dumpPython on the config that gets created (e.g.,
> file.write(process.dumpPython) )



well,
```
runTheMatrix --dryRun
```
does not create the python files at all, so I cannot easily dump them :-(

.A

**Comment by davidlange6** [comment]
its runTheMatrix --no_exec if I remember correctly

**Comment by mark-grimes** [comment]
Still digesting this, but on the point of dumping config files:
`--no_exec` doesn't work and I don't see anything similar in the runTheMatrix source file (I could have missed it though).
I've always used

    runTheMatrix -l <something> --dryRun
    cd <something>*
    sed s:">":"--no_exec #": cmdLog | sh

(adds `--no_exec`, comments out the redirections then executes the cmsDriver command)
I guess if you're doing everything (no `-l`) you'll need something like

    runTheMatrix --dryRun
    ls | awk '{print "cd "$0"; sed s:\">\":\"--no_exec #\": cmdLog | sh; cd .."}' | sh

Bit of a ballache but it works.  Then when I was having to check for the eras I used

    ls */*.py | awk -F "\\\\.py" '{print "edmConfigDump "$0" > "$1"_dump.py"}' | sh

(The `\\\\` stops awk interpreting `.` as "any character" and matching against e.g. `_pythia8`)
Note that any print statements in the configuration tree are also added to the output and can stop the dumped files from being python parseable, but this works well for a simple diff.

**Comment by fwyzard** [comment]
Hi Mark,

On 10 February 2015 at 22:04, Mark Grimes <notifications@github.com> wrote:

> --no_exec doesn't work and I don't see anything similar in the
> runTheMatrix source file (I could have missed it though).
>

I've used

runTheMatrix.py --command=--no_exec

to create all .py files, and then

find * -name '*.py' | xargs -n1 -i -P16 sh -c 'edmConfigDump "$1" >
"$1"_dump.py' -- {}

to expand them all.

-P16 runs 16 jobs in parallel, which is handy if you have enough cores.


The diff of all expanded dumps with and without my changes counted over 30k
lines, so I've given up checking it after a while, and encounterng only
expected (or anyway plausible) changes.

.A

**Comment by Martin-Grunewald** [comment]
Whoa hold on. ```--no_exec``` must be kept working, we run our tests this way!

**Comment by Martin-Grunewald** [comment]
This PR now has merge conflicts....

**Comment by fwyzard** [comment]
@Martin-Grunewald I did not touch the `--no_exec` functionality, I'm just reporting how to use it with `runTheMatrix.py`

**Comment by fwyzard** [comment]
rebased on top of the latest CMSSW_7_4_X

**Comment by slava77** [comment]
I'm usually using "-j 0" (I think this way of making --no_exec predates
appearance of --command)
does this still work?



On 2/11/15 2:42 AM, Andrea Bocci wrote:
> @Martin-Grunewald <https://github.com/Martin-Grunewald> I did not touch
> the |--no_exec| functionality, I'm just reporting how to use it with
> |runTheMatrix.py|
> 
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/7579#issuecomment-73848652>.
> 

-- 
***********************************************************************
 Vyacheslav (Slava) Krutelyov
 TAMU: Physics Dept Texas A&M MS4242, College Station, TX 77843-4242
 CERN: 42-R-027
 AIM/Skype: siava16              googleTalk: slava77@gmail.com
 (630) 291-5128  Cell (US)       +41 76 275 7116 Cell (CERN)

-----------------------------------------------------------------------

**Comment by davidlange6** [comment]
@cmsbuild, please test

**Comment by davidlange6** [comment]
bypassing since we already agreed on this PR (and its comparisons are all ok)

**Comment by franzoni** [comment]
hi physics-week => missed the train 
:-(

I believe that this: digiPremixUp2015Defaults50ns 
should also receive the 50ns specific customisation.
@mdhildreth can you also cross check, please ?

@fwyzard : what effect do we expect on HLT performance in pre7; it would be good if you / Darren / Roberta could spell out the impact to be explained to the validators

These are less important aspects:
step3Up2015Hal  => should be customised as 25 ns ?
HARVESTGEN => no need of any customisation

**Comment by fwyzard** [comment]
On 12 February 2015 at 19:09, Giovanni Franzoni <notifications@github.com>
wrote:

> I believe that this: digiPremixUp2015Defaults50ns
> should also receive the 50ns specific customisation.
> @mdhildreth <https://github.com/mdhildreth> can you also cross check,
> please ?
>
I guess so, from the name, but what is that ?
Something in the matrix ?

@fwyzard <https://github.com/fwyzard> : what effect do we expect on HLT
> performance in pre7; it would be good if you / Darren / Roberta could spell
> out the impact to be explained to the validators
>
No L1 or HLT changes to the 25ns amples from this PR (though of course
there may be changes from the other L1 or HLT-ralated PR's).

The 50ns samples should be properly reverted to the Legacy L1 emulator, so
they should observe changes to most calorimetric related L1 quantities
(Jets, MET, E/Gamma, Taus). Do we actually have 50ns relvals ?

The HI samples should properly use the HI-specific settings for the Stage 1
L1 emulator, but I do not know what that means in practice (I would guess a
different treatment of the pileup, but I do not really know).

These are less important aspects:
> step3Up2015Hal => should be customised as 25 ns ?
> HARVESTGEN => no need of any customisation
>
.A

**Comment by franzoni** [comment]
see #7826 for follow up about premixing workflows

**Pull request merged**